File: Extra_things/Stack and queue/stack.py




#! IN NEED OF TESTING AND REFACTORING



class Stack:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return not self.items

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if self.is_empty():
            raise IndexError("Cannot pop from an empty stack")
        return self.items.pop()

    def peek(self):
        if self.is_empty():
            raise IndexError("Cannot peek from an empty stack")
        return self.items[-1]

    def size(self):
        return len(self.items)

File: Extra_things/Stack and queue/stack.cpp




//! IN NEED OF TESTING AND REFACTORING



template <typename T>
class stack
{
    struct Node
    {
        T data;
        Node* next;
        Node(T data) : data(data), next(nullptr) {}
    };

    Node* top;
    int size;

public:
    stack() : top(nullptr), size(0) {}

    ~stack()
    {
        while (top)
        {
            Node* temp = top;
            top = top->next;
            delete temp;
        }
    }

    void push(T data)
    {
        Node* newNode = new Node(data);
        newNode->next = top;
        top = newNode;
        size++;
    }

    T pop()
    {
        if (!top) throw std::runtime_error("Stack is empty");
        T data = top->data;
        Node* temp = top;
        top = top->next;
        delete temp;
        size--;
        return data;
    }

    int getSize() const { return size; }

    bool empty() const { return !top; }
};
